### DISCUSSION TITLE
Changes Required for a TPU Pod

### DISCUSSION THREAD

**Discussion by agemagician** [google/flax#2017] [Q&A] [answered]
Hello,

Is there any kind of changes that need to be done on the Flax/Jax code to make it run on a TPU pod rather than a single TPU node ?

for example the T5 example on HuggingFace:
https://github.com/huggingface/transformers/blob/main/examples/flax/language-modeling/run_t5_mlm_flax.py
Runs normally on a single TPU Pod, but if we need to run it on a TPU Pod, do we need to make any kind of changes to the code?

or we just need to run the same script across all workers and it will communicate with each other:
`gcloud alpha compute tpus tpu-vm ssh tpu-vm --zone=us-central2-b --worker=all --command "python myscript.py"`

I am trying to run HuggingFace code on TPU V4-64 pod and your reply will be really helpful :)

**Comment by agemagician** [comment]
The reason for this question is that several users who have access to TPU V4 have a big issue.

The training script for hugging face runs at the same speed and global batch size in a single node or a pod.
For example. running flax t5 in a TPU V4-64 pod reduces the per device batch size to 4 and makes the global batch size to 128, while In a single node it increases to 32 per device and makes the global batch size to 128.
The training speed is exactly the same in a single node or a pod.

**Comment by jheek** [comment]
The first thing to do here would be to add a log message to your script printing the output of `jax.device_count()` or even better `jax.devices()` so you can make sure all devices are discovered by the TPU client. But this does sound like a bug where the different TPU hosts cannot find each other. It's best to report this as an issue in the JAX repo

**Reply by dptam** [reply]
Hi @agemagician, I was also running into an issue with jax on TPU V4, but when I ran `jax.devices`, it only showed the CPU. I can run it on a TPU V3 and it will show all the devices. I was wondering if you did anything specific to set it up? Thanks

**Reply by agemagician** [reply]
> Hi @agemagician, I was also running into an issue with jax on TPU V4, but when I ran `jax.devices`, it only showed the CPU. I can run it on a TPU V3 and it will show all the devices. I was wondering if you did anything specific to set it up? Thanks

I don't have this problem, I just removed jax and installed the latest version and it worked fine.

**Comment by agemagician** [comment, marked as answer]
It actually discovers all devices and it prints them correctly. 
Ok, then I will report it in the Jax repo.
Thanks.

**Reply by jheek** [reply]
Hmm in that case could you try to increase the global batch size by a factor 8 on the pod experiment?

**Reply by agemagician** [reply]
I did that and also doesn't work :(

I am not sure, but it got stuck at the sharding step for data.

Maybe there is a bug in "shard" function in flax for a tpu pod :
https://github.com/google/flax/blob/b60f7f45b90f8fc42a88b1639c9cc88a40b298d3/flax/training/common_utils.py#L23

It seems that the shard function divide the data per local device which in this case will always be 4 for TPU-V4. While in Huggingface example, they send to the shard function the global batch (per_device_batch * jax.device_count()).

Do you think this function should divide the batch using the device_count :

```
def shard(xs):
  device_count = jax.device_count()
  return jax.tree_map(
      lambda x: x.reshape((device_count, -1) + x.shape[1:]), xs)
```

**Reply by jheek** [reply]
No because a host should only provide data for the local devices. Are you perhaps creating a batch of size = global_batch_size instead of global_batch_size / num_hosts?

**Reply by agemagician** [reply]
Yep, this is the way they did it in HF.
```
# Gather the indexes for creating the batch and do a training step
for step, batch_idx in enumerate(tqdm(train_batch_idx, desc="Training...", position=1)):
    samples = [tokenized_datasets["train"][int(idx)] for idx in batch_idx]
    model_inputs = data_collator(samples)

    # Model forward
    model_inputs = shard(model_inputs.data)
```

The "model_inputs" input, in this case, is the global batch size (per_device_batch * jax.device_count()).

I think there are 2 solutions here:
1. Divide the model_inputs per host and send the correct part to each host.
2. Make a new function in flax that takes a global batch and shard it into each host, then shard it again into local tpus.

**Reply by jheek** [reply]
There is a new experimental feature called GlobalDeviceArray's that makes this easier by giving you a view into the "global" arrays rather than than just the local part. For now though the way JAX works is that each host is responsible for feeding the local devices so the input script should produce a batch that is either `per_device_batch * jax.local_device_count()` or `global_batch_size / jax.process_count()`. The former is better in a sense because it supports a setup where not every host as the same number of GPUs/TPUs (but that's rare).

**Reply by agemagician** [reply]
Yep, that makes a lot of sense.
Thanks a lot :)

**Reply by agemagician** [reply]
I will ping the HF team to be aware of this issue.

**Reply by congyingxia** [reply]
Hi @agemagician, do you have example code that can be shared for solving this issue?  I met the same problem when I want to adapt the code on TPU-V3-8 to TPU-V3-32. I've read your discussions, but still not sure how to fix it. Do you use the GlobalDeviceArray? Thank you in advance!

**Reply by agemagician** [reply]
> Hi @agemagician, do you have example code that can be shared for solving this issue? I met the same problem when I want to adapt the code on TPU-V3-8 to TPU-V3-32. I've read your discussions, but still not sure how to fix it. Do you use the GlobalDeviceArray? Thank you in advance!

Hi @congyingxia ,
You can check the required change in my pull request which was already integrated into HF for T5 model:
https://github.com/huggingface/transformers/pull/16527/files

**Reply by congyingxia** [reply]
Thank you so much!

**Comment by agemagician** [comment]
Thanks again for your effort.
